peps/pep-0733.rst

```diff
@@ -0,0 +1,634 @@
+In particular, implementation details such as the fields of C structs
+can be exposed
+[`Issue 22 <https://github.com/capi-workgroup/problems/issues/22>`__
+and :pep:`620`].
+This can make API evolution very difficult, in particular when it
+occurs in the stable ABI as in the case of ``ob_refcnt`` and ``ob_type``,
+which are accessed via the reference counting macros
+[`Issue 45 <https://github.com/capi-workgroup/problems/issues/45>`__].
+
+We identified a deeper issue in relation to the way that reference
+counting is exposed. The way that C extensions are required to
+manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
+specific to CPython's memory model, and is hard for alternative
+Python implementations to emulate.
+[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].
+
+Another set of problems arises from the fact that a ``PyObject*`` is
+exposed in the C API as an actual pointer rather than a handle. The
+address of an object serves as its ID and is used for comparison,
+and this complicates matters for alternative Python implementations
+that move objects during GC
+[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].
+
+A separate issue is that object references are opaque to the runtime,
+discoverable only through calls to ``tp_traverse``/``tp_clear``,
+which have their own purposes. If there was a way for the runtime to
+know the structure of the object graph, and keep up with changes in it,
+this would make it possible for alternative implementations to implement
+different memory management schemes
+[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].
+
+Object Reference Management
+---------------------------
+
+There are C API functions that return borrowed references, and
+functions that steal references to arguments, but there isn't a
+naming convention that makes this obvious, so this is error prone
+[`Issue 8 <https://github.com/capi-workgroup/problems/issues/8>`__
+and `Issue 52 <https://github.com/capi-workgroup/problems/issues/52>`__].
+The terminology used to describe these situations in the documentation
+can also be improved
+[`Issue 11 <https://github.com/capi-workgroup/problems/issues/11>`__].
+
+A more radical change is necessary in the case of functions that
+return borrowed references (such as ``PyList_GetItem``)
+[`Issue 5 <https://github.com/capi-workgroup/problems/issues/5>`__ and
+`Issue 21 <https://github.com/capi-workgroup/problems/issues/21>`__]
+or pointers to parts of the internal structure of an object
+(such as ``PyBytes_AsString``)
+[`Issue 57 <https://github.com/capi-workgroup/problems/issues/57>`__].
+In both cases, the reference/pointer is valid for as long as the
+owning object is alive, but this time is hard to reason about. Such
+functions should not exist in the API without a mechanism that can
+make them safe.
+
+For containers, the API is currently missing bulk operations on the
```

Comment:
Nit: I would call them "batched" operations and also suggest that the batch size and starting point is configurable by the function call.

Reply:
Can you suggest an edit?

Reply:
I think the current wording is fine. This paragraph describes a problem; it doesn't have to outline the exact API for the solution.

Reply:
Fair enough :-)